Support syslog(2): add syscall(2), logger integration and tests

TankTechnology opened the pull request:
- Add `sys_syslog` module with circular 64 KB kernel log buffer and `dmesg_restrict`.
- Extend `AsterLogger` to push logs into the buffer and honor console log levels.
- Register callback in syscall init, dispatch on x86/RISCV (`SYS_SYSLOG`).

cchanging reviewed (commented):
Thank you for further enriching Asterinas's log system! This is great work.

I took a look at this PR, and the code quality seems solid. However, while reviewing, I noticed a few main points that might need clarification:  

The first is about `register_syslog_callback`. The `aster_logger` crate is meant to provide a concrete logger implementation, while the more internal `ostd` already has `ostd::logger::inject_logger`. I think you could directly implement the logging logic within the `aster_logger` crate instead.  

The second is about `log_level`. Since `sys_syslog` has the ability to control `log_level`, you introduced a global variable to track the log level and attempted to use it to filter console output. However, I’m uncertain whether your current `log_level` implementation will work as intended.

Currently, Asterinas primarily uses the `log` crate’s macros (`debug!`, `warn!`, etc.), and the `log` crate itself has its own log level. These macros already include logic to check the `log` crate’s level before call `log` method.  If Asterinas doesn’t specify a `LOG_LEVEL` parameter, the `log` crate defaults to `error` level. In that case, even if you set `sys_log`’s level to `all`, the `debug!` and `warn!` messages won’t be output or stored in the buffer—because the `log` crate’s level filtering prevents them from reaching the actual logging function.  Perhaps we should integrate with the `log` crate’s functionality to achieve this properly?

cchanging reviewed (commented):
Thank you for continuously improving this PR! I feel the current implementation has brought us one step closer to merging.

A few reminders regarding our commit practices in Asterinas: We typically maintain independent, clear commits - if an earlier commit has issues, we generally amend it directly rather than adding new fix commits. Therefore, I suggest squashing the latest two commits into their corresponding previous ones.

When you're satisfied with the changes, please rebase to the latest main branch to avoid merge conflicts.

tatetian reviewed (commented):
Good job on adding the `syslog` system call.

TankTechnology commented on kernel/comps/logger/src/aster_logger.rs:
```diff
@@ -1,7 +1,263 @@
+const KERN_DEBUG: i32 = 7; // Debug messages
+
+/// Efficient circular kernel log buffer
+struct KernelLogBuffer {
```
This has been addressed: the ring-buffer was extracted into `kernel/libs/ring-buffer` as a standalone crate, merged via #2870. The klog buffer now uses `RingBuffer<u8>` from that crate, backed by `Segment<()>` instead of `Vec`.

TankTechnology commented on kernel/comps/logger/src/aster_logger.rs:
```diff
@@ -1,7 +1,263 @@
+const KERN_DEBUG: i32 = 7; // Debug messages
+
+/// Efficient circular kernel log buffer
+struct KernelLogBuffer {
```
The ring-buffer has been extracted into `kernel/libs/ring-buffer` (merged via #2870) as `RingBuffer<T>` + `Producer`/`Consumer` for `T: Pod`, with push/pop/push_slice/pop_slice operations. Unit tests for the ring-buffer crate can be added as a follow-up PR.

TankTechnology commented on kernel/comps/logger/src/aster_logger.rs:
```diff
@@ -1,7 +1,263 @@
+    /// Add a log message to the buffer
+    fn append(&mut self, data: &[u8]) {
+        // Limit message size to prevent overwhelming the buffer
+        let max_msg_size = self.buffer.len() / 4; // Max 25% of buffer per message
```
The 25% `max_msg_size` magic number has been removed. The current implementation uses a fixed `FORMAT_BUF_CAPACITY` of 512 bytes, independent of the buffer size. Messages exceeding 512 bytes are silently truncated.

TankTechnology commented on kernel/comps/logger/src/aster_logger.rs:
```diff
@@ -1,7 +1,263 @@
+        // Limit message size to prevent overwhelming the buffer
+        let max_msg_size = self.buffer.len() / 4; // Max 25% of buffer per message
+        let data = if data.len() > max_msg_size {
+            &data[..max_msg_size]
```
Each `append_log` call writes one `writeln!`-delimited record (ending with `
`). The `SYSLOG_ACTION_READ` operation reads raw bytes — this matches Linux `syslog(2)` behavior, which does not guarantee record boundaries. If `/dev/kmsg` or `/proc/kmsg` are added later, per-record headers can be introduced then.

TankTechnology commented on kernel/comps/logger/src/aster_logger.rs:
```diff
@@ -1,7 +1,263 @@
+
+        self.total_written += data.len();
+
+        for &byte in data {
```
Fixed. The `append()` method now uses `buf.push_slice(bytes)` for bulk writing, and `read()`/`read_all()` use `copy_from()` which reads in chunks of `COPY_CHUNK` (512 bytes).

TankTechnology commented on kernel/comps/logger/src/aster_logger.rs:
```diff
@@ -1,7 +1,263 @@
+        let unread_count = self.unread_count();
+        let to_read = core::cmp::min(buf.len(), unread_count);
+
+        for (i, item) in buf.iter_mut().enumerate().take(to_read) {
```
Fixed. `read()` and `read_all()` now use `copy_from()` which reads in chunks of `COPY_CHUNK` (512 bytes) via `Segment::read_slice`.

TankTechnology commented on kernel/comps/logger/src/aster_logger.rs:
```diff
@@ -15,7 +271,19 @@ impl log::Log for AsterLogger {
-        print_logs(record, timestamp);
+
+        // Always add to kernel log buffer for syslog functionality
+        add_to_kernel_log(record.level(), &record.args().to_string());
```
Fixed. The logger now uses `FixedBuf`, a stack-allocated formatter backed by a fixed-size `[u8; 512]` array. No `format!`, `to_string`, or heap allocation occurs in the logging path.

TankTechnology commented on kernel/comps/logger/src/aster_logger.rs:
```diff
@@ -1,7 +1,263 @@
+        _ => log::LevelFilter::Trace,  // All messages
+    };
+    
+    log::set_max_level(log_level);
```
Fixed. `set_console_log_level` no longer calls `log::set_max_level`. It only controls console output via the `should_print` check, which compares `(level as u8) < (console_level as u8)`.

TankTechnology commented on kernel/comps/logger/src/aster_logger.rs:
```diff
@@ -1,7 +1,263 @@
+}
+
+/// Functions for syslog syscall integration
+pub fn read_kernel_log_destructive(buf: &mut [u8]) -> usize {
```
Fixed. Functions like `klog_read`, `klog_read_all`, `klog_wait_nonempty`, etc. remain in `aster_logger::klog` as the buffer API, but the syscall dispatch logic (`sys_syslog`, `read_destructive`, `read_all`, `ensure_cap`) has been moved to `kernel/src/syscall/syslog.rs`.

TankTechnology commented on kernel/comps/logger/src/aster_logger.rs:
```diff
@@ -1,7 +1,263 @@
+}
+
+/// Update the console log level from external components
+pub fn set_console_log_level(level: i32) {
```
Fixed. Added `SysLogAction` enum with `TryFromInt` for syscall action dispatch, and `LinuxConsoleLogLevel` enum with `from_raw()` constructor and `should_print()` method. These replace the raw constants and helper functions from the original implementation.

TankTechnology commented on kernel/src/syscall/syslog.rs:
```diff
@@ -0,0 +1,178 @@
+const SYSLOG_ACTION_CONSOLE_ON: i32 = 7;
+const SYSLOG_ACTION_CONSOLE_LEVEL: i32 = 8;
+const SYSLOG_ACTION_SIZE_UNREAD: i32 = 9;
+const SYSLOG_ACTION_SIZE_BUFFER: i32 = 10;
```
Done. The syscall actions are now modeled as a `SysLogAction` enum with `TryFromInt` derivation, following the same pattern as `MMapType` in `syscall/mmap.rs`.

TankTechnology commented on kernel/comps/logger/src/aster_logger.rs:
```diff
@@ -1,7 +1,263 @@
+}
+
+/// Global kernel log buffer (lazily initialized)
+static KERNEL_LOG_BUFFER: SpinLock<Option<KernelLogBuffer>> = SpinLock::new(None);
```
The current implementation uses `SpinLock<Option<Arc<KernelLog>>>` instead of `Once` because OSTD does not provide a `Once` type. The `init_klog()` function ensures the buffer is initialized before first use, and the lazy initialization in `klog()` serves as a fallback.

TankTechnology commented on kernel/comps/logger/src/aster_logger.rs:
```diff
@@ -1,7 +1,263 @@
+use ostd::{sync::SpinLock, timer::Jiffies};
+
+/// Console log level (default is 7 - show all messages)
+static CONSOLE_LOG_LEVEL: SpinLock<i32> = SpinLock::new(7);
```
Done. The console level is now stored as an `AtomicU8` with `Ordering::SeqCst`, which avoids the overhead of a `SpinLock`.

TankTechnology commented on kernel/src/syscall/arch/x86.rs:
```diff
@@ -266,6 +267,7 @@ impl_syscall_nums_and_dispatch_fn! {
     SYS_GETTIMEOFDAY = 96      => sys_gettimeofday(args[..1]);
     SYS_GETRLIMIT = 97         => sys_getrlimit(args[..2]);
     SYS_GETRUSAGE = 98         => sys_getrusage(args[..2]);
+    SYS_SYSLOG = 103           => sys_syslog(args[..3]);
```
Fixed. `SYS_SYSLOG = 116` has been added to the generic syscall table in `kernel/src/syscall/arch/generic.rs`, which covers RISC-V and other non-x86 architectures.

TankTechnology commented on test/src/apps/scripts/run_general_test.sh:
```diff
@@ -8,8 +8,10 @@ SCRIPT_DIR=/test
-./test_epoll_pwait.sh
-./cgroup.sh
+# ./test_epoll_pwait.sh
+# ./cgroup.sh
```
After rebasing onto the latest main branch, the test scripts (`test_epoll_pwait.sh` and `cgroup.sh`) are no longer part of the diff. The test infrastructure was refactored upstream, and the syslog tests will be adapted to the new `test/initramfs/src/regression/` framework in a follow-up.

TankTechnology commented:
## Summary of latest push

The branch has been rebased onto the latest `main` and cleaned up. Here's what changed:

### Structural changes
- **Rebased onto upstream `main`**: The ring-buffer extraction (#2870) has been merged separately, so this PR no longer carries those commits.
- **Squashed from 14 commits to 2**: As requested, fix commits have been folded into their corresponding logical commits. The two commits are:
  1. `Add klog buffer and wire sys_syslog syscall` — core implementation
  2. `Add /proc/sys/kernel/dmesg_restrict sysctl` — procfs interface

### All review feedback addressed
| # | Reviewer | Comment | Resolution |
|---|---|---|---|
| 1 | tatetian | Reuse existing ring buffer | `RingBuffer<T>` from `kernel/libs/ring-buffer` (#2870) |
| 2 | tatetian | Extract CircularBuffer + unit tests | Crate extracted via #2870; unit tests as follow-up |
| 3 | tatetian | 25% magic number | Removed; uses fixed 512-byte `FORMAT_BUF_CAPACITY` |
| 4 | tatetian | Record boundaries | `\n` delimiter matches Linux `syslog(2)` behavior |
| 5/6 | tatetian | Byte-level read/write | Bulk `push_slice`/`copy_from` in `COPY_CHUNK` (512) |
| 7 | tatetian | No heap allocation in logger | `FixedBuf` stack-allocated formatter |
| 8 | tatetian | Don't set `log` crate level | `set_max_level` removed; `should_print` uses `(level as u8) < (self as u8)` |
| 9 | tatetian | Extract syscall code | Moved to `kernel/src/syscall/syslog.rs` |
| 10 | tatetian | Types over functions | `SysLogAction` enum + `LinuxConsoleLogLevel` type |
| - | cchanging | Squash + rebase | 14 commits → 2, based on latest main |
| - | cchanging | `SysLogAction` enum | Done |
| - | cchanging | `AtomicU8` for console level | Done |
| - | Copilot | RISC-V syscall registration | Added `SYS_SYSLOG = 116` to generic table |
| - | Copilot | Disabled test scripts | No longer in diff after rebase |

### Scope
- **11 files changed, +755/-21** (down from 24 files, +1867/-441)
- Compiles cleanly (only pre-existing `VDSO_LIBRARY_DIR` build issue remains)
- Tests will be adapted to the new `test/initramfs/src/regression/` framework in a follow-up

TankTechnology committed:
Add klog buffer and wire sys_syslog syscall

Implement the syslog(2) system call with a lock-free ring-buffer-based
kernel log buffer. The implementation includes:

- A 64 KB circular kernel log buffer (RingBuffer<u8>) with console log
  level filtering, dmesg_restrict security flag, and blocking read
- Integration with AsterLogger to capture log messages into the buffer
  with timestamp and prefix formatting, using only stack allocation
- SysLogAction enum and LinuxConsoleLogLevel type for type-safe syslog
  dispatch
- Registration of SYS_SYSLOG for x86_64 (103) and generic (116, covers
  RISC-V and other architectures)

tatetian commented:
@TankTechnology Thanks for the revision. It looks that you have been assisted with AI. That's fine. But I need your confirmation that you have (1) conducted a thorough self-review (mandatory) and (2) used the [`pr-review-skill`](https://github.com/tatetian/asterinas/blob/pr-review-skill/.agents/skills/pr-review/SKILL.md) (recommended).

Contributors outnumbers reviewers. So it is highly desirable for contributors to polish their code as much as possible before requesting for manual review. I marked this PR as "Draft". Once you complete the manual and AI-assisted self-review, feel free to mark this PR "Ready for Review".

TankTechnology committed:
Fix klog IRQ safety, API shape, and Linux syslog semantics

- Replace SpinLock<Option<Arc<KernelLog>>> with Once<KernelLog> to remove
  per-log-call lock contention and Arc::clone overhead
- Use LocalIrqDisabled guard on all log-path spinlocks to prevent deadlock
  when IRQ handlers log while a CPU holds a buffer lock
- Snapshot buffer tail in read_all for cross-chunk consistency within a
  single syslog(2) call
- Fix size_unread to account for clear_tail so it returns 0 after Clear
- Fix SizeUnread to respect dmesg_restrict instead of unconditionally
  requiring CAP_SYSLOG, matching Linux behavior
- Fix ConsoleLevel to reject len outside 1..=8 (EINVAL), matching Linux
- Split set_console_level into set_console_level and disable_console to
  remove the boolean argument per project coding guidelines
- Add module-level documentation with lock ordering and invariants
- Reset clear_tail alongside buf.clear() in oversized-message truncation

TankTechnology committed:
Fix klog code organization and dmesg_restrict write_at return value

- Move COPY_CHUNK and FixedBuf closer to top for top-down readability
- Return input_len instead of read_len in write_at per convention

TankTechnology reviewed (commented):
## Self-review complete, ready for review

I have conducted a thorough self-review and addressed the following issues:

### IRQ safety (blocker)
All spinlocks on the log emission path (`buffer`, `clear_tail`, `saved_console_level`) now use `LocalIrqDisabled` guard, preventing deadlock when IRQ handlers log while a CPU holds a log lock.

### API shape (significant)
Replaced `SpinLock<Option<Arc<KernelLog>>>` + 16 free-function wrappers with `Once<KernelLog>` + `pub fn klog() -> &'static KernelLog`. Eliminates per-log-call lock contention and `Arc::clone` overhead. Callers use `aster_logger::klog().method()` consistently.

### read_all snapshot consistency (blocker)
Added `snapshot_tail()` — the syscall handler captures the tail once, and all chunks within one syscall use the same snapshot, preventing data skipping/duplication.

### size_unread respects clear_tail (blocker)
Now computes `tail - max(head, clear_tail)`, returns 0 after `SYSLOG_ACTION_CLEAR`.

### Linux syslog semantics (blocker)
- **SizeUnread (action 9)**: Now respects `dmesg_restrict` instead of unconditionally requiring CAP_SYSLOG.
- **ConsoleLevel (action 8)**: Validates len in 1..=8, returning EINVAL for 0.

### Coding guideline fixes
- **no-bool-args**: Split `set_console_level` into `set_console_level` and `disable_console`.
- **SAFETY comment**: Replaced misleading `// SAFETY:` on safe code with a plain comment.
- **Module doc**: Added `//!` preamble with lock ordering and invariants.
- **Top-down reading**: Moved `COPY_CHUNK` and `FixedBuf` near the top.
- **Import style**: Uses `aster_logger::klog().method()` per AGENTS.md.

### Minor fixes
- `dmesg_restrict` write_at returns `input_len` instead of `read_len`.
- `bump_clear_tail` takes `&RingBuffer` instead of `&mut`.
- Append truncation resets `clear_tail` alongside `buf.clear()`.
- `DMESG_RESTRICT_MAX_INPUT` moved to top of file.

### Verification
- `make kernel` builds successfully
- `make test` passes all tests
- `make ktest` passes (235 tests)
- 21 C regression tests covering all SysLogAction codes pass in QEMU

### Tests (follow-up PR on `syslog-tests` branch)
- 21 `#[cfg(ktest)]` unit tests for ring buffer, clear_tail, console levels
- 21 C regression tests under `test/initramfs/src/regression/process/syslog/`

TankTechnology commented:
Closing this PR since #3288 has taken over the syslog(2) implementation and further improves on this work.

Thanks everyone for the reviews and guidance. I will follow the progress in #3288.
